### ISSUE TITLE
Animation performance

### ISSUE THREAD

**Issue by trvrb** [nextstrain/auspice#367]
Check if there's low hanging fruit to improve animation performance. The 100ms tick can probably be improved for Zika. @colinmegill: How was 100ms tick chosen? Ideally, we want this to tick as fast as render works but no faster.

**Comment by colinmegill** [comment]
@trvrb am going to do a branch for requestAnimationFrame now that we know we want animations with fixed times, I would expect this to be an improvement.

**Comment by trvrb** [comment]
Excellent.

**Comment by jameshadfield** [comment]
Cool. For what it's worth, I think the fixed times should be faster. `flu` with `slow` takes 4min to animate on my laptop!

**Comment by jameshadfield** [comment]
Flu animation render stack time (i.e. the code executed during each `setInterval`) is around 560ms for me (about 1/3 of the way through the tree). Of this, `getLatLongs` is running twice and taking 260ms + 175ms. `updateStylesAndAttrs` (tree) is taking 80ms. Those 3 calls are 92% of the stack time.

**Comment by trvrb** [comment]
I'm getting 220ms frames (4 FPS) for default Ebola animation on production. This looks to be about 100 ms for tree and 100 ms for map. `getLatLongs` is indeed running twice (taking 20ms each time). `updateVisibility` is a separate call taking 70ms.

![profile](https://user-images.githubusercontent.com/1176109/27980979-6d2d9cf2-6339-11e7-8915-2fe1713d4512.png)

This more what I'm expecting. 3y flu is about as big as Ebola and I'm getting 410ms frames on production. `getLatLongs` is taking 200 ms (twice). I thought that this might be because this was "country" geo resolution, but we have the same issue with "region" geo resolution when flu has 10 regions and Ebola has 50. Or I guess it has 30 given it's tripled. Okay. This makes sense. If flu is not tripled then performance is similar to Ebola with `getLatLongs` taking 20ms instead of 200 ms. Tripling demes results in 9x more lines. So 20ms vs 200ms makes sense.

Moving forward I see two obvious ideas:

* Don't fully triple map. We need a "west" copy of Japan, but not a "west" copy of Europe. Need a half copy on the left and a half copy on the right giving 2x demes. I don't know how difficult this would be to implement.
* Avoid the duplicate `geLatLongs` call.

There may be internal logic that can get cleaned up as well.

**Comment by trvrb** [comment]
Commit e405742711de697a67bee0033af74c018b384e1d fixes the duplicate `latLongs` call. This was happening because `componentDidUpdate` fires twice after updating `dateMin` or `dateMax`. One firing because of `props.controls` and then a second firing once tree finishes updating it's visibility. I'll merge this commit but we can keep working on the issue.

**Comment by trvrb** [comment]
Digging into internals of `getLatLongs`. The thing that takes up all the time is this loop:

https://github.com/nextstrain/auspice/blob/master/src/util/mapHelpersLatLong.js#L130

For each of 3 offsets x each of ~500 transmission events x 3 offsets = ~4500 calculations each cycle for flu. There should be ways to both reduce the number of calculations and/or to pre-compute. These are basically the same calculations being done each cycle. This is where we will get most improvement in particular for flu, but others s well.

**Comment by jameshadfield** [comment]
After merge of PR #377 the code execution of map is much improved:
```
setInterval stack start
map CWRP starting
map CWRP finished
map CDU starting
map CDU finished
map CWRP starting
map CWRP finished
map CDU starting
maybeUpdateDemesAndTransmissions running
getLatLongs running
map CDU finished
setInterval stack finished
```

As noted by @trvrb, `changeDateFilter` results in 2 actions (see #348) - it may be possible to combine these into 1, however the above console.logs show that all the `maybe` functions in map are exiting appropriately so it's probably not going to make a dramatic difference.

**Comment by jameshadfield** [comment]
Current status of flu @ 006b8924385f2f4d35944d32713e5ae26040ff94

![image](https://user-images.githubusercontent.com/8350992/28043810-86aa2dde-6588-11e7-8f0a-2cbf8f7e6b46.png)
![image](https://user-images.githubusercontent.com/8350992/28043833-ab45e494-6588-11e7-8631-b86036492898.png)

Putting console log statements in the `setInterval` block led me to believe that (e.g.) maybeUpdateDemesAndTransmissions was only running once, however the above screenshots indicate a different story. There appear to be 7 dispatches, but this is misleading as thunks wrap code in the dispatch function - i.e. there are not 7 actions dispatched.

* Tick compute time: 475ms
* `setState` on some component: 16ms (unsure which component)
* `calcVisibility`: 7ms (inside thunk)
* (tree) `updateStylesAndAttrs`: 14ms (via `componentWillRecieveProps`)
* (tree) `updateStylesAndAttrs` (again!): 121ms (via `componentWillRecieveProps`)
  * `calcStrokeCols`: 74ms
* (tree) `updateStylesAndAttrs` (a third time!): 31ms (via `componentWillRecieveProps`)
  * `mapToScreen`: 25ms
* `maybeUpdateDemesAndTransmissions`: 16ms (via `componentDidUpdate`)
* `maybeUpdateDemesAndTransmissions` (again): 111ms (via `componentDidUpdate`)
  * `getLatLongs`: 34ms
  * `updateVisibility`: 77ms (incl Bezier calcs)
* `maybeUpdateDemesAndTransmissions` (a third time): 138ms (via `componentDidUpdate`)
  * `updateVisibility`: 138ms (incl Bezier calcs)

**Comment by trvrb** [comment]
This is excellent. Thanks James. I believe `maybeUpdateDemesAndTransmissions` was getting called multiple times for flu but not for Zika.

**Comment by jameshadfield** [comment]
Summary of things to do now:
- [ ] stop unnecessary tree colorBy changes #384
- [ ] single dispatch during tick #385
- [ ] precompute leaflet lat long objects
- [ ] precompute bezier curve arrays
- [ ] don't update SVG transmissions not in view

**Comment by jameshadfield** [comment]
`Recalculate Style (CSSPropertOperations.js?...)` takes around 30ms on flu - anyone know what this is from? Radium?

**Comment by jameshadfield** [comment]
Profile as of `master` 1f07afcc7a8cf7bd7d970c0ddf5af873f38007a4 (single dispatch during tick, no unnecessary tree colorBy updates)

![image](https://user-images.githubusercontent.com/8350992/28077382-7adbfae4-6616-11e7-88b8-15b2b18c7685.png)

A single action is always being dispatched through the reducer (confirmed, to the best of my ability, via console logs -- see https://github.com/nextstrain/auspice/pull/389#issue-242077111). Despite this `maybeDemesAndTransmissions` is called between 1 and 7 (!) times (it varies between ticks). I'll leave it to @colinmegill to track down the cause, but I'm very interested in why it's doing this. Perhaps the conditionals can be improved to bail out appropriately?

Testing at http://localhost:4000/flu/h3n2/ha/6y?c=region&dmin=2013-09-20&r=region, **ticks take between 288 and 308ms** plus around 35ms to recalculate CSS styles.
* Almost the entire time is spent inside (tree) `updateStylesAndAttrs`, (map) `maybeUpdateDemesAndTransmissions` and the the recalculation of CSS styles.
* (tree) `updateStylesAndAttrs` is consistently running once and takes 21-32ms,
  * `phyloTree.mapToScreen` is around 50-80% of this
* `maybeUpdateDemesAndTransmissions` runs up to 7 times and takes up to 250ms
  * `updateVisibility` is the largest contributor to this

**Comment by jameshadfield** [comment]
I've been investigating React Peformance Tools - https://facebook.github.io/react/docs/perf.html - and have implemented on the branch `react-perf` to display the time taken in each component during an animation tick. Note: `React` and some other libraries needed updating, so you'll have to run `npm install`

Time spent during each animation tick (on flu, link above):
![image](https://user-images.githubusercontent.com/8350992/28083726-3895932e-662b-11e7-92c9-d15eefa16a50.png)

Not sure if I can get it to break `Map` apart more, but this is a quantitative way by which we may examine performance.

**Comment by trvrb** [comment]
Cool. Thanks for adding this James. Do you have a TLDR version of getting this readout?

**Comment by jameshadfield** [comment]
```
# be prepared for npm issues
git checkout react-perf
npm install
node dev-server.js
# play animation and look at console
```

**Comment by jameshadfield** [comment]
As of 69158dabe85a02702bf48aa9907d5429e4962fa3 the animation speed is much improved - now it really looks as if the tree is the bottleneck

**Comment by trvrb** [comment]
If we debounce `mapToScreen` (#387) and fix Bezier redrawing (#381) I think we'll be close to 50ms for Ebola.

**Comment by trvrb** [comment]
Commit 4e1119b9cdbb26c95ce1b551f8e875f649a85924 has `mapToScreen` debounced (#387) and Bezier fixed (#381). This is showing hugely improved performance. Original map PR (at 31516f224392881825d5b9382b719a4f7cf4ec90) had Ebola with 115ms render (plus 15ms repaint):

<img width="1112" alt="tick-old" src="https://user-images.githubusercontent.com/1176109/29245103-a3bc42e4-7f83-11e7-8812-7bb6556240c1.png">

Current animation has Ebola with 15ms (plus 15ms repaint):

<img width="1012" alt="tick-new" src="https://user-images.githubusercontent.com/1176109/29245106-c5f9e3e8-7f83-11e7-9950-4e79ff8c0c3a.png">

So, we just need to clean up the last remaining issues in fix Bezier (#381) so it can be merged. And then separately we need to switch to `requestAnimationFrame` (#404). With #381 and #404 in place this issue will be resolved and animation can launch.

**Comment by trvrb** [comment]
I'm going to close this now. With #381 and #387 in place animation performance is what it should be. I'm considering animation timing (#404) a separate issue.
